cmd/release/runner.go

```diff
@@ -0,0 +1,231 @@
+}
+
+// replaceWorkInProgressVersionWithRelease replaces work in progress version with the release that we want to publish in the source code.
+func (r *runner) replaceReleaseVersionWithNextWorkInProgress(file string, worktree *git.Worktree) (plumbing.Hash, error) {
+	err := r.replaceVersionInFile(file, r.flag.CurrentVersion, r.flag.NextPatchWorkInProgressVersion)
+	if err != nil {
+		return plumbing.Hash{}, microerror.Mask(err)
+	}
+
+	commit, err := r.addAndCommitChanges(file, worktree, r.flag.Author, fmt.Sprintf("bump new work in progress version to %s", r.flag.NextPatchWorkInProgressVersion))
+	if err != nil {
+		return plumbing.Hash{}, microerror.Mask(err)
+	}
+
+	return commit, err
+}
+
+func (r *runner) addAndCommitChanges(file string, worktree *git.Worktree, author *object.Signature, commitMessage string) (plumbing.Hash, error) {
+	_, err := worktree.Add(strings.TrimPrefix(file, r.flag.RepositoryPath+"/"))
+	if err != nil {
+		return plumbing.Hash{}, microerror.Mask(err)
+	}
+
+	commit, err := worktree.Commit(commitMessage, &git.CommitOptions{
+		Author: author,
+	})
+
+	return commit, microerror.Mask(err)
+}
+
+func (r *runner) replaceVersionInFile(file, search, replaceWith string) error {
+	f, err := ioutil.ReadFile(file)
+	if err != nil {
+		return microerror.Mask(err)
+	}
+	filecontents := string(f)
+
+	if !strings.Contains(filecontents, search) {
+		return microerror.Maskf(NoVersionFoundInFileError, "No version was found in %s", file)
+	}
+
+	updatedFileContents := []byte(strings.Replace(filecontents, search, replaceWith, 1))
+	err = ioutil.WriteFile(file, updatedFileContents, 0)
+	if err != nil {
+		return microerror.Mask(err)
+	}
+
+	return nil
+}
+
+func (r *runner) addReleaseToChangelog(file string, worktree *git.Worktree) (plumbing.Hash, error) {
+	search := "## [Unreleased]"
+	replaceWith := fmt.Sprintf("## [Unreleased]\n\n## [%s] %s", r.flag.CurrentVersion, time.Now().Format("2006-01-02"))
+	f, err := ioutil.ReadFile(file)
+	if err != nil {
+		return plumbing.Hash{}, microerror.Mask(err)
+	}
+	filecontents := string(f)
+
+	if !strings.Contains(filecontents, search) {
+		return plumbing.Hash{}, microerror.Maskf(NoUnreleasedWorkFoundInChangelogError, "No unreleased work was found in %s", file)
+	}
+
+	updatedFileContents := []byte(strings.Replace(filecontents, search, replaceWith, 1))
+	err = ioutil.WriteFile(file, updatedFileContents, 0)
+	if err != nil {
+		return plumbing.Hash{}, microerror.Mask(err)
+	}
+
+	// Change [Unreleased] link
+	m1 := regexp.MustCompile(`(\[Unreleased]:)(.*)(v[0-9]+\.[0-9]+\.[0-9]+)`)
+	updatedFileContents = []byte(m1.ReplaceAllString(string(updatedFileContents), fmt.Sprintf("$1${2}%s$5", r.flag.TagName)))
+	err = ioutil.WriteFile(file, updatedFileContents, 0)
+	if err != nil {
+		return plumbing.Hash{}, microerror.Mask(err)
+	}
+
+	// Change new tag's link
+	taglink := fmt.Sprintf("https://github.com/%s/%s/releases/tag/%s", r.flag.Organization, r.flag.RepositoryName, r.flag.TagName)
```

Comment:
It isn't correct. It should be `compare` URL. See example below. Note first, last and all in the middle.

```
[Unreleased]: https://github.com/giantswarm/architect-orb/compare/v0.8.9...HEAD

[0.8.9]: https://github.com/giantswarm/architect-orb/compare/v0.8.8...v0.8.9
[0.8.8]: https://github.com/giantswarm/architect-orb/compare/v0.8.7...v0.8.8
[0.8.7]: https://github.com/giantswarm/architect-orb/compare/v0.8.6...v0.8.7
[0.8.6]: https://github.com/giantswarm/architect-orb/compare/v0.8.5...v0.8.6
[0.8.5]: https://github.com/giantswarm/architect-orb/compare/v0.8.4...v0.8.5
[0.8.4]: https://github.com/giantswarm/architect-orb/compare/v0.8.3...v0.8.4
[0.8.3]: https://github.com/giantswarm/architect-orb/compare/v0.8.2...v0.8.3
[0.8.2]: https://github.com/giantswarm/architect-orb/compare/v0.8.1...v0.8.2
[0.8.1]: https://github.com/giantswarm/architect-orb/compare/v0.8.0...v0.8.1
[0.8.0]: https://github.com/giantswarm/architect-orb/compare/v0.7.0...v0.8.0
[0.7.0]: https://github.com/giantswarm/architect-orb/compare/v0.6.0...v0.7.0
[0.6.0]: https://github.com/giantswarm/architect-orb/compare/v0.5.3...v0.6.0
[0.5.3]: https://github.com/giantswarm/architect-orb/compare/v0.5.2...v0.5.3
[0.5.2]: https://github.com/giantswarm/architect-orb/compare/v0.5.1...v0.5.2
[0.5.1]: https://github.com/giantswarm/architect-orb/compare/v0.5.0...v0.5.1
[0.5.0]: https://github.com/giantswarm/architect-orb/compare/v0.4.5...v0.5.0
[0.4.5]: https://github.com/giantswarm/architect-orb/compare/v0.4.4...v0.4.5
[0.4.4]: https://github.com/giantswarm/architect-orb/compare/v0.4.3...v0.4.4
[0.4.3]: https://github.com/giantswarm/architect-orb/compare/v0.4.2...v0.4.3
[0.4.2]: https://github.com/giantswarm/architect-orb/compare/v0.4.1...v0.4.2
[0.4.1]: https://github.com/giantswarm/architect-orb/compare/v0.4.0...v0.4.1
[0.4.0]: https://github.com/giantswarm/architect-orb/compare/v0.3.0...v0.4.0
[0.3.0]: https://github.com/giantswarm/architect-orb/compare/v0.2.0...v0.3.0
[0.2.0]: https://github.com/giantswarm/architect-orb/compare/v0.1.2...v0.2.0
[0.1.2]: https://github.com/giantswarm/architect-orb/compare/v0.1.1...v0.1.2
[0.1.1]: https://github.com/giantswarm/architect-orb/compare/v0.1.0...v0.1.1

[0.1.0]: https://github.com/giantswarm/architect-orb/releases/tag/v0.1.0
```